src/transformers/models/mplugdocowl/modeling_mplugdocowl.py

```diff
@@ -0,0 +1,1038 @@
+            `(batch_size, num_heads, encoder_sequence_length, embed_size_per_head)`.
+
+            Contains pre-computed hidden-states (key and values in the self-attention blocks and in the cross-attention
+            blocks) that can be used (see `past_key_values` input) to speed up sequential decoding.
+
+            If `past_key_values` are used, the user can optionally input only the last `decoder_input_ids` (those that
+            don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of all
+            `decoder_input_ids` of shape `(batch_size, sequence_length)`.
+        inputs_embeds (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`, *optional*):
+            Optionally, instead of passing `input_ids` you can choose to directly pass an embedded representation. This
+            is useful if you want more control over how to convert `input_ids` indices into associated vectors than the
+            model's internal embedding lookup matrix.
+        vision_feature_layer (`int`, *optional*, defaults to -2):
+            The index of the layer to select the vision feature.
+        vision_feature_select_strategy (`str`, *optional*, defaults to `"default"`):
+            The feature selection strategy used to select the vision feature from the vision backbone.
+            Can be one of `"default"` or `"full"`.
+        use_cache (`bool`, *optional*):
+            If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding (see
+            `past_key_values`).
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    """The MPLUGDOCOWL model which consists of a vision backbone and a language model.""",
+    MPLUGDOCOWL_START_DOCSTRING,
+)
+# Copied from transformers.models.llava.modeling_llava.LlavaForConditionalGeneration with LLAVA->MPLUGDOCOWL,Llava->MPLUGDocOwl,llava->mplugdocowl
+class MPLUGDocOwlForConditionalGeneration(MPLUGDocOwlPreTrainedModel):
+    def __init__(self, config: MPLUGDocOwlConfig):
+        super().__init__(config)
+        self.vision_tower = AutoModel.from_config(config.vision_config)
+
+        self.multi_modal_projector = MPLUGDocOwlMultiModalProjector(config)
+        self.vocab_size = config.text_config.vocab_size
+        #initialize LlamaAttention
+        #replace_llama_modality_adaptive()
+        transformers.models.llama.modeling_llama.LlamaAttention = MultiwayAttention
```

Comment:
I am not sure whether this changes the model architecture, because it still doesn't have the modules that I need. Like Multiway Attention or new Decoder.

Reply:
Hi, was seeing this PR so might help here :)

Usually we avoid using inheritance or classes from different models. Each model is typically implemented in a single modeling_xxx.py script which is self-contained and does not depend on any other modeling files.

This means that we'll need to define a MPlugDowOwlMultiwayAttention class, a MPlugDowOwlTextDecoderLayer class, and so on.

One can leverage `#Copied from` statements above a class or method in case a class or method is the same as another one, e.g. `MistralAttention` copies from `LlamaAttention` since they're the same.

Reply:
Hello Niels! @molbap provided a similar feedback:) Yes, I just created a separate file for the language model and defined separate classes. This indeed helps! Thank you!